### PR TITLE
Upgrade azure-sdk-versions

### PR DESCRIPTION
**Pull request by unaizorrilla**
**What this PR does / why we need it**:
Upgrade Azure SDK and include new TokenCredential overload #606

**Which issue(s) this PR fixes**:

Please reference the issue this PR will close: #606

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [ x] Code compiles correctly
- [ ] Created/updated tests
- [ x] Unit tests passing
- [ x] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### CODE REVIEW

**Review by CarlosLanderas — APPROVED** [review]

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 1]
```diff
@@ -11,71 +12,68 @@ namespace HealthChecks.AzureKeyVault
...
             {
-                using (var client = CreateClient())
+                foreach (var secret in _options.Secrets)
                 {
-                    foreach (var secret in _options.Secrets)
-                    {
-                        await client.GetSecretAsync(_options.KeyVaultUrlBase, secret, cancellationToken);
-                    }
+                    var secretClient = CreateSecretClient();
+                    await secretClient.GetSecretAsync(_options.KeyVaultUrlBase, secret, cancellationToken);
```
Comment (heaths) [inline src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs:31]: The vault URI is no longer passed into the method, but only once into the constructor for all clients. Used this way, you're passing the `KeyVaultUrlBase` as the name of the secret and the `secret` is being passed as the optional version. Same is true for keys and certificates.

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 2]
```diff
@@ -11,71 +12,68 @@ namespace HealthChecks.AzureKeyVault
...
-            if (_options.UseManagedServiceIdentity)
-            {
-                var azureServiceTokenProvider = new AzureServiceTokenProvider();
-                return new KeyVaultClient(new KeyVaultClient.AuthenticationCallback(azureServiceTokenProvider.KeyVaultTokenCallback));
-            }
-            else
-            {
-                return new KeyVaultClient(GetToken);
-            }
+            return new KeyClient(_keyVaultUri, _azureCredential);
```
Comment (heaths) [inline src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs:66]: As mentioned above, you should create these once.

File: src/HealthChecks.AzureStorage/AzureBlobStorageHealthCheck.cs [hunk 3]
```diff
@@ -9,20 +10,32 @@ namespace HealthChecks.AzureStorage
...
+            _azureCredential = credential ?? throw new ArgumentNullException(nameof(credential));
+            _containerName = containerName;
+        }
+
         public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
         {
             try
             {
-                var blobServiceClient = new BlobServiceClient(_connectionString);
+                var blobServiceClient = GetBlobServiceClient();
```
Comment (heaths) [inline src/HealthChecks.AzureStorage/AzureBlobStorageHealthCheck.cs:36]: This client should also be stored. Probably just do this and the other clients in the constructor. They are pretty lightweight and, if arguments are all valid, really shouldn't fail. Or at least do a lazy-init.

File: src/HealthChecks.AzureStorage/DependencyInjection/AzureStorageHealthCheckBuilderExtensions.cs [hunk 4]
```diff
@@ -82,5 +84,30 @@ public static IHealthChecksBuilder AddAzureQueueStorage(this IHealthChecksBuilde
                tags,
                timeout));
         }
+
+        /// <summary>
+        /// Add a health check for Azure Queue Storage.
+        /// </summary>
+        /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+        /// <param name="queueServiceUri">The Azure Queue service Uri (https://myaccount.blob.core.windows.net)). </param>
+        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzure Credential or other kind of credentials,you can read more on https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md. </param>
```
Comment (heaths) [inline src/HealthChecks.AzureStorage/DependencyInjection/AzureStorageHealthCheckBuilderExtensions.cs:93]: You can use `<see href="url-here"/>` to encode the URL so it's clickable in Visual Studio and rendered in any HTML docs (depending on the generator, but most seem to do it).

**Review by heaths — CHANGES_REQUESTED** [review]
I highly recommend creating the clients as singletons (at least for a given URI), but mainly the calls to methods like `GetSecret` need to be changed so that you don't pass the URIs in there as well. At least for Key Vault, that won't get you the certificate, key, or secret you want (it would actually 404 since some characters in a URI are invalid name characters and couldn't exist).

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs (hunk 1 above)
Comment (unaizorrilla) [reply src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs:31]: True, my fault on migrate from older version, I try to fix asap!

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 5]
```diff
@@ -11,71 +12,68 @@ namespace HealthChecks.AzureKeyVault
...
             try
             {
-                using (var client = CreateClient())
+                foreach (var secret in _options.Secrets)
                 {
-                    foreach (var secret in _options.Secrets)
-                    {
-                        await client.GetSecretAsync(_options.KeyVaultUrlBase, secret, cancellationToken);
-                    }
+                    var secretClient = CreateSecretClient();
```
[the remark this answers is not in the document]
Comment (unaizorrilla) [reply src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs:30]: Well, this is the same with other healthchecks, we need to do some more code here! dev's can register different healchecks for different Key vaults .
I try to add the same workaround that use on other healtchecks like this 
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/e424418c0ca34fa3d81777a1268f347da6f100f5/src/HealthChecks.AzureServiceBus/AzureServiceBusTopicHealthCheck.cs#L13

sound good?

File: src/HealthChecks.AzureStorage/AzureBlobStorageHealthCheck.cs (hunk 3 above)
Comment (unaizorrilla) [reply src/HealthChecks.AzureStorage/AzureBlobStorageHealthCheck.cs:36]: Wait for response on my comment above about singleton healthchecks

File: src/HealthChecks.AzureStorage/DependencyInjection/AzureStorageHealthCheckBuilderExtensions.cs (hunk 4 above)
Comment (unaizorrilla) [reply src/HealthChecks.AzureStorage/DependencyInjection/AzureStorageHealthCheckBuilderExtensions.cs:93]: True, also my fault.. I write the code in hurry :-(

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs (hunk 2 above)
Comment (unaizorrilla) [reply src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs:66]: waiting for your response on my comment above

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (heaths) [reply src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs:30]: That is a good idea, yes.

File: src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs [hunk 6]
```diff
@@ -1,42 +1,42 @@
...
+{
+    public static class AzureKeyVaultHealthChecksBuilderExtensions
+    {
+        const string KEYVAULT_NAME = "azurekeyvault";
+
+        /// <summary>
+        /// Add a health check for Azure Key Vault. Default behaviour is using Managed Service Identity, to use Client Secrets call UseClientSecrets in setup action
+        /// </summary>
+        /// <param name="keyVaultServiceUri">The AzureKeyVault service uri.</param>
+        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzureCredential or other kind of TokenCredential,you can read more on <see cref="https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md"/>. </param>
```
Comment (heaths) [inline src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs:17]: Not a `cref`, but an `href`. `cref` will try to link to a type.

File: src/HealthChecks.AzureStorage/DependencyInjection/AzureStorageHealthCheckBuilderExtensions.cs [hunk 7]
```diff
@@ -89,8 +89,8 @@ public static IHealthChecksBuilder AddAzureQueueStorage(this IHealthChecksBuilde
         /// Add a health check for Azure Queue Storage.
         /// </summary>
         /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
-        /// <param name="queueServiceUri">The Azure Queue service Uri (https://myaccount.blob.core.windows.net)). </param>
-        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzure Credential or other kind of credentials,you can read more on https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md. </param>
+        /// <param name="queueServiceUri">The Azure Queue service Uri like <see cref="https://myaccount.blob.core.windows.net"/>. </param>
+        /// <param name="credential">The TokenCredentail to use, you can use Azure.Identity with DefaultAzureCredential or other kind of TokenCredential,you can read more on <see cref="https://github.com/Azure/azure-sdk-for-net/blob/Azure.Identity_1.2.2/sdk/identity/Azure.Identity/README.md"/>. </param>
```
Comment (heaths) [inline src/HealthChecks.AzureStorage/DependencyInjection/AzureStorageHealthCheckBuilderExtensions.cs:93]: Same here: `href`.

**Review by heaths — APPROVED** [review]
Looks good. Thank you!

**Comment by CarlosLanderas** [comment]
Thanks for the feedback @heaths. You rock!

**Comment by naile** [comment]
Just updated to the latest version here and noticed Table Storage support has been removed in this PR. Was this intentional? (can't see a mention of that in the PR description).

**Comment by unaizorrilla** [comment]
Yeap!!

Azure.data.tables is not published yet, Microsoft confirm the publishing date on this mont! We need this because Azure.Storage no longer support TableClieny

**Pull request merged**